internal/internal_poller_autoscaler.go

```diff
@@ -0,0 +1,168 @@
+// IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+// FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+// AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+// LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+// OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+// THE SOFTWARE.
+
+package internal
+
+import (
+	"context"
+	"errors"
+	"github.com/marusama/semaphore/v2"
+	"go.uber.org/atomic"
+	"go.uber.org/cadence/internal/common/autoscaler"
+	"go.uber.org/zap"
+	"time"
+)
+
+// defaultPollerScalerCooldownInSeconds
+const (
+	defaultPollerScalerCooldownInSeconds = 120
+)
+
+var (
+	_ autoscaler.AutoScaler = (*pollerAutoScaler)(nil)
+	_ autoscaler.Estimator  = (*pollerUsageEstimator)(nil)
+)
+
+type (
+	pollerAutoScaler struct {
+		pollerUsageEstimator
+
+		isDryRun     bool
+		cooldownTime time.Duration
+		logger       *zap.Logger
+		sem          semaphore.Semaphore // resizable semaphore to control number of concurrent pollers
+		ctx          context.Context
+		cancel       context.CancelFunc
+		recommender  autoscaler.Recommender
+	}
+
+	pollerUsageEstimator struct {
+		// This single atomic variable stores two variables:
+		// left 32 bits is noTaskCounts, right 32 bits is taskCounts.
+		// This avoids unnecessary usage of CompareAndSwap
+		atomicBits *atomic.Uint64
+	}
+)
+
+func newPollerScaler(
+	initialPollerCount int,
+	minPollerCount int,
+	maxPollerCount int,
+	targetUtilizationInMilli uint64,
+	isDryRun bool,
+	cooldownTime time.Duration,
+	logger *zap.Logger) *pollerAutoScaler {
+	ctx, cancel := context.WithCancel(context.Background())
+	return &pollerAutoScaler{
+		isDryRun:             isDryRun,
+		cooldownTime:         cooldownTime,
+		logger:               logger,
+		sem:                  semaphore.New(initialPollerCount),
+		ctx:                  ctx,
+		cancel:               cancel,
+		pollerUsageEstimator: pollerUsageEstimator{atomicBits: atomic.NewUint64(0)},
+		recommender: autoscaler.NewLinearRecommender(
+			autoscaler.Resource(minPollerCount),
+			autoscaler.Resource(maxPollerCount),
+			autoscaler.Usages{
+				autoscaler.PollerUtilizationRate: autoscaler.UsageInMilli(targetUtilizationInMilli),
+			},
+		),
+	}
+}
+
+// Acquire concurrent poll quota
+func (p *pollerAutoScaler) Acquire(resource autoscaler.Resource) error {
+	return p.sem.Acquire(p.ctx, int(resource))
+}
+
+// Release concurrent poll quota
+func (p *pollerAutoScaler) Release(resource autoscaler.Resource) {
+	p.sem.Release(int(resource))
+}
+
+// GetCurrent poll quota
+func (p *pollerAutoScaler) GetCurrent() autoscaler.Resource {
+	return autoscaler.Resource(p.sem.GetLimit())
+}
+
+// Start an auto-scaler go routine and returns a done to stop it
+func (p *pollerAutoScaler) Start() autoscaler.DoneFunc {
+	logger := p.logger.Sugar()
+	go func() {
+		for {
+			select {
+			case <-p.ctx.Done():
```

Comment:
This should return when cancelled so the goroutine doesn't leak?

Reply:
good catch!